### ISSUE TITLE
Cwd::realpath fails on unsearchable directory (AIX)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9513]
Migrated from [rt.perl.org#59662](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=59662) (status was 'resolved')

Searchable as RT59662$

**Comment by p5pRT** [comment]
### From John.Wiersba@thomsonreuters.com

#### Created by john.wiersba@thomsonreuters.com


 
1\) Compile the following program as \./a\.out&#8203;:
 
/\* C program to demonstrate that system\-supplied realpath\(\) works\. \*/
\#include \<stdio\.h>
\#include \<stdlib\.h>
int main \( void \)
\{
&nbsp;   char buf\[1000\+1\];
&nbsp;   realpath\("testdir"\, buf\);
&nbsp;   printf\("\<%s>\\n"\, buf\);
\}
 
2\) Run the following script&#8203;:
 
\#\!/bin/ksh
mkdir testdir
chmod 444 testdir
echo "from perl&#8203;:"
perl \-we '
&nbsp;   use strict;
&nbsp;   use Cwd \(\);
&nbsp;   print "\<"\, Cwd&#8203;::realpath\("testdir"\)\, ">\\n";
'
echo "from a\.out&#8203;:"
\./a\.out
rmdir testdir
 
3\) See the following output&#8203;:
 
from perl&#8203;:
Use of uninitialized value in print at \-e line 4\.
\<>
from a\.out&#8203;:
\</home/jrw32982/junk/testdir>
 
4\) Analysis&#8203;:
 
Perl's realpath\(\) is aliased to abs\_path\(\) which is aliased to
\_perl\_abs\_path\(\)\.  When \_perl\_abs\_path\(ARG\) sees that ARG is a
directory\, it
tries to opendir\("ARG/\.\."\) which fails because ARG is not searchable\.
The
system\-supplied realpath\(\) does not try to opendir\("ARG/\.\."\) and does
not
fail\.
 

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.8.2:
 
Configured by root at Thu Jan 26 14:26:31 CST 2006.
 
Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=aix, osvers=5.2.0.0, archname=aix-thread-multi
    uname='aix animals 2 5 000709cf4c00 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc_r', ccflags ='-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE
-qmaxmem=16384 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT
-q32 -D_LARGE_FILES -qlonglong',
    optimize='-O',
    cppflags='-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE
-qmaxmem=16384 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT
-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=16384
-qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT -q32
-D_LARGE_FILES -qlonglong -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE
-qmaxmem=16384 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT
-q32 -D_LARGE_FILES -qlonglong -D_ALL_SOURCE -D_ANSI_C_SOURCE
-D_POSIX_SOURCE -qmaxmem=16384 -qnoansialias -DUSE_NATIVE_DLOPEN
-DNEED_PTHREAD_INIT -q32 -D_LARGE_FILES -qlonglong -D_ALL_SOURCE
-D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=16384 -qnoansialias
-DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT -q32 -D_LARGE_FILES -qlonglong
-D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=16384
-qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT -q32
-D_LARGE_FILES -qlonglong -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE
-qmaxmem=16384 -qnoansialias -DUSE_NATIVE_DLOPEN -DNEED_PTHREAD_INIT
-q32 -D_LARGE_FILES -qlonglong'
    ccversion='5.0.2.0', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld', ldflags =' -brtl -b32 -bmaxdata:0x80000000'
    libpth=/lib /usr/lib /usr/ccs/lib
    libs=-lbind -lnsl -ldbm -ldl -lld -lm -lpthreads -lc_r -lcrypt -lbsd
-lPW
    perllibs=-lbind -lnsl -ldl -lld -lm -lpthreads -lc_r -lcrypt -lbsd
-lPW
    libc=/lib/libc.a, so=a, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_aix.xs, dlext=so, d_dlsymun=undef,
ccdlflags='-bE:/usr/opt/perl5/lib/5.8.2/aix-thread-multi/CORE/perl.exp
-bE:/usr/opt/perl5/lib/5.8.2/aix-thread-multi/CORE/perl.exp
-bE:/usr/opt/perl5/lib/5.8.2/aix-thread-multi/CORE/perl.exp
-bE:/usr/opt/perl5/lib/5.8.2/aix-thread-multi/CORE/perl.exp'
    cccdlflags=' ', lddlflags='-bhalt:4 -bM:SRE -bI:$(PERL_INC)/perl.exp
-bE:$(BASEEXT).exp -bnoentry -lpthreads -lc_r'
 
Locally applied patches:
    SUIDPERLIO1 - fix PERLIO_DEBUG buffer overflow (CAN-2005-0156)
    SPRINTF0 - fixes for sprintf formatting issues - CVE-2005-3962
 

@INC for perl v5.8.2:
    /usr/opt/perl5/lib/5.8.2/aix-thread-multi
    /usr/opt/perl5/lib/5.8.2
    /usr/opt/perl5/lib/site_perl/5.8.2/aix-thread-multi
    /usr/opt/perl5/lib/site_perl/5.8.2
    /usr/opt/perl5/lib/site_perl
    .
 

Environment for perl v5.8.2:
    HOME=/home/jrw32982
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=C
    LC__FASTMSG=true
    LD_LIBRARY_PATH (unset)
    LIBPATH (unset)
    LOGDIR (unset)
 
PATH=/home/jrw32982/perl/bin:/medstat/devl/tools/common/scripts:/home/jr
w32982/sh:/home/jrw32982/bin:/bin:/usr/bin:/usr/sbin:/usr/bin/X11:/usr/l
ocal/bin:/usr/local/sbin:/etc:/sbin:/usr/ccs/bin:/usr/ucb:/usr/lbin:/med
stat/devl/tools/AIX_5.3/bin:.
    PERL_BADLANG (unset)
    SHELL=/usr/bin/ksh


```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

On Mon Oct 06 13&#8203;:18&#8203;:14 2008\, John\.Wiersba@&#8203;thomsonreuters\.com wrote&#8203;:
> This is a bug report for perl from john\.wiersba@&#8203;thomsonreuters\.com\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
>  
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>  
> 1\) Compile the following program as \./a\.out&#8203;:
>  
> /\* C program to demonstrate that system\-supplied realpath\(\) works\. \*/
> \#include \<stdio\.h>
> \#include \<stdlib\.h>
> int main \( void \)
> \{
>    char buf\[1000\+1\];
>    realpath\("testdir"\, buf\);
>    printf\("\<%s>\\n"\, buf\);
> \}
>  
> 2\) Run the following script&#8203;:
>  
> \#\!/bin/ksh
> mkdir testdir
> chmod 444 testdir
> echo "from perl&#8203;:"
> perl \-we '
>    use strict;
>    use Cwd \(\);
>    print "\<"\, Cwd&#8203;::realpath\("testdir"\)\, ">\\n";
> '
> echo "from a\.out&#8203;:"
> \./a\.out
> rmdir testdir
>  
> 3\) See the following output&#8203;:
>  
> from perl&#8203;:
> Use of uninitialized value in print at \-e line 4\.
> \<>
> from a\.out&#8203;:
> \</home/jrw32982/junk/testdir>
>  
> 4\) Analysis&#8203;:
>  
> Perl's realpath\(\) is aliased to abs\_path\(\) which is aliased to
> \_perl\_abs\_path\(\)\.  When \_perl\_abs\_path\(ARG\) sees that ARG is a
> directory\, it
> tries to opendir\("ARG/\.\."\) which fails because ARG is not searchable\.
> The
> system\-supplied realpath\(\) does not try to opendir\("ARG/\.\."\) and does
> not
> fail\.
>  

This seems to be fixed in recent Cwd's\.\.\.

riker&#8203;:perl\-current tanya$ mkdir tempdir
riker&#8203;:perl\-current tanya$ chmod 444 tempdir
riker&#8203;:perl\-current tanya$ \./perl \-Ilib \-MCwd \-Mstrict \-wle'print "\<"\, Cwd&#8203;::realpath\("testdir
"\)\, ">"'
\</Users/tanya/perl\-current/testdir>


I'd suggest upgrading to a recent Cwd from CPAN\.

Steve Peters

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From John.Wiersba@thomsonreuters.com

Steve\,

I think you're probably right that this is fixed in at least some
situations in later releases\.

I'm following up because there is a bug in the latest Cwd\.pm \(version
3\.29\)\.  Line 546 calls File&#8203;::Spec\->rel2abs\(\)\, but File&#8203;::Spec has not yet
been loaded\.  The fix is similar to line 526&#8203;:  add&#8203;:

&nbsp;   require File&#8203;::Spec;

right before calling File&#8203;::Spec\->rel2abs\(\) on line 546\.

I was able to see this bug in action by copying \*only\* the Cwd\.pm from
the latest module release for Cwd \(from PathTools\-3\.29\.tar\.gz\) to the
current directory and then adding 

&nbsp;   use lib "\."

to the perl part of my test script\.  This caused the following error&#8203;:

Can't locate object method "rel2abs" via package "File&#8203;::Spec" at Cwd\.pm
line 546\.

\#\!/bin/ksh
mkdir testdir
chmod 444 testdir
echo "from perl&#8203;:"
perl \-we '
&nbsp;   use strict;
&nbsp;   use lib "\.";
&nbsp;   use Cwd \(\);
&nbsp;   print "\<"\, Cwd&#8203;::realpath\("testdir"\)\, ">\\n";
'
echo "from a\.out&#8203;:"
\./a\.out
rmdir testdir

\(The reason I copied only the Cwd\.pm file from the PathTools package is
that the latest version of Cwd\.pm that I have installed on any of my
systems is 3\.12 and I don't know how to easily build a local copy of Cwd
from the PathTools package that won't need to be installed in the global
system directory and therefore interfere with or overwrite the standard
system installed one\)\.

\-\- John Wiersba

> \-\-\-\-\-Original Message\-\-\-\-\-
> From&#8203;: Steve Peters via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\] 
> Sent&#8203;: Monday\, December 01\, 2008 11&#8203;:58 PM
> To&#8203;: Wiersba\, John \(TH USA\)
> Subject&#8203;: \[perl \#59662\] \[RESOLVED\] Cwd&#8203;::realpath fails on 
> unsearchable directory \(AIX\) 
> 
> According to our records\, your request regarding 
>   "Cwd&#8203;::realpath fails on unsearchable directory \(AIX\)" 
> has been resolved\. 
> 
> If you have any further questions or concerns\, please respond 
> to this message\.
> 
> For other topics\, please create a new ticket\.
> 
> Please don't feel obligated to say "Thanks" or "Kudos" or "I 
> owe you a beer" \-\- if you respond to this message it will 
> reopen the ticket\. If you must\, please send email directly to 
> the person who handled your ticket\, and not to the tracking system\.
> 
> \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=59662 >
> 
>
